pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol

```diff
@@ -59,73 +76,351 @@ contract LiquidityBootstrappingPool is BaseWeightedPool, ReentrancyGuard {
-            if (normalizedWeight > maxNormalizedWeight) {
-                maxWeightTokenIndex = i;
-                maxNormalizedWeight = normalizedWeight;
-            }
         }
         // Ensure that the normalized weights sum to ONE
         _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
 
-        // Initial value; can change later if weights change
-        _maxWeightTokenIndex = maxWeightTokenIndex;
+        // Write initial pool state - all zeros except for the start weights
+        _poolState = poolState;
+
+        // If false, the sale will start in a paused state (prevents front-running the unpause transaction)
+        swapEnabled = swapEnabledOnStart;
+
+        emit PublicSwapSet(swapEnabledOnStart);
+    }
+
+    // External functions
+
+    /**
+     * @dev Return start time, end time, and endWeights as an array
+     */
+    function getGradualUpdateParams()
+        external
+        view
+        returns (
+            uint256 startTime,
+            uint256 endTime,
+            uint256[] memory endWeights
+        )
+    {
+        // Load current pool state from storage
+        bytes32 poolState = _poolState;
+
+        startTime = poolState.decodeUint32(_START_TIME_OFFSET);
+        endTime = poolState.decodeUint32(_END_TIME_OFFSET);
+        uint256 totalTokens = _getTotalTokens();
+        endWeights = new uint256[](totalTokens);
+
+        // prettier-ignore
+        {
+            if (totalTokens > 0) { endWeights[0] = poolState.uncompress16(_END_WEIGHT_OFFSET); }
+            if (totalTokens > 1) { endWeights[1] = poolState.uncompress16(_END_WEIGHT_OFFSET + 16); }
+            if (totalTokens > 2) { endWeights[2] = poolState.uncompress16(_END_WEIGHT_OFFSET + 32); }
+            if (totalTokens > 3) { endWeights[3] = poolState.uncompress16(_END_WEIGHT_OFFSET + 48); }
+        }
+    }
+
+    /**
+     * @dev Can pause/unpause trading
+     */
+    function setPublicSwap(bool publicSwap) external authenticate whenNotPaused nonReentrant {
+        swapEnabled = publicSwap;
+
+        emit PublicSwapSet(publicSwap);
     }
 
+    /**
+     * @dev Schedule a gradual weight change, from the current weights to the given endWeights,
+     * over startTime to endTime
+     */
+    function updateWeightsGradually(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory endWeights
+    ) external authenticate whenNotPaused nonReentrant {
+        // solhint-disable-next-line not-rely-on-time
+        uint256 currentTime = block.timestamp;
+
+        _require(currentTime < endTime, Errors.GRADUAL_UPDATE_TIME_TRAVEL);
+
+        // Must specify normalized weights for all tokens
+        uint256 totalTokens = _getTotalTokens();
+        InputHelpers.ensureInputLengthMatch(totalTokens, endWeights.length);
+
+        // If the start time is in the past, "fast forward" to start now
+        // This prevents circumventing the minimum weight change duration
```

Comment:
We seem to allow zero duration weight changes so this comment doesn't look correct. Was the minimum removed?

Reply:
Leftover comment, thanks! Discussed this with Nico, who made an analogy with Stable Pools, and said a minimum duration might give a false sense of security. First, it can be set to zero anyway (and in fact, most V1 smart pools left it at the 2 minute GUI default, effectively zero). They all could have changed the weights drastically more or less instantly = arbitrary price manipulation, and the users didn't care (and nobody actually did that).

It could possibly be effective if you made it very long, or there were a hard-coded minimum duration (>> 0). That's all up for discussion - it does require trust now, as it does on V1. But for now, there is no minimum duration, so I'll remove that stale comment.

One improvement in V2 over V1 is you can now see what the end weights are, and the start/end times.

Reply:
There is still a reason to do this - preventing sudden discontinuities. I updated the comment.